### PR TITLE
Remove traps

### PR DESCRIPTION
**Pull request by miramor**

### CODE REVIEW

File: lib/mods/theme/edit/p_info.txt [hunk 1]
```diff
@@ -182,8 +180,8 @@ C:a:k:=0:=0:Axe-mastery
 C:a:k:=0:=0:Hafted-mastery
 C:a:k:=0:=0:Polearm-mastery
 C:a:k:+1000:-200:Archery
-C:a:k:+1000:+900:Sneakiness
-C:a:k:+1000:+900:Disarming
+
+
```
Comment (BardurArantsson) [inline lib/mods/theme/edit/p_info.txt]: Spurious newlines left here

File: lib/mods/theme/edit/p_info.txt [hunk 2]
```diff
@@ -1039,8 +1001,8 @@ C:k:+0:+200:Axe-mastery
 C:k:+0:+200:Hafted-mastery
 C:k:+0:+200:Polearm-mastery
 C:k:+1000:+600:Archery
-C:k:+1000:+900:Sneakiness
-C:k:+1000:+900:Disarming
+
```
Comment (BardurArantsson) [inline lib/mods/theme/edit/p_info.txt]: Newlines

File: lib/mods/theme/edit/p_info.txt [hunk 3]
```diff
@@ -1536,11 +1466,10 @@ R:C:Loremaster
 R:C:Mage
 R:C:Pacifist
 R:C:Priest
-R:k:+600:+000:Disarming
 R:k:+000:+000:Magic-Device
 R:k:+5000:+000:Spirituality
 R:k:-16000:+000:Stealth
-R:k:+3000:+000:Sneakiness
+
```
Comment (BardurArantsson) [inline lib/mods/theme/edit/p_info.txt]: newline

File: lib/mods/theme/edit/p_info.txt [hunk 4]
```diff
@@ -1563,11 +1492,10 @@ R:C:Mage
 R:C:Pacifist
 R:C:Priest
 #R:G:EVIL
-R:k:+1000:+000:Disarming
 R:k:+500:+000:Magic-Device
 R:k:-4000:+000:Spirituality
 R:k:-10000:+000:Stealth
-R:k:+000:+000:Sneakiness
+
```
Comment (BardurArantsson) [inline lib/mods/theme/edit/p_info.txt]: newline

File: lib/mods/theme/edit/p_info.txt [hunk 5]
```diff
@@ -2322,8 +2230,8 @@ S:F:SPEED
 S:Z:poison dart
 R:k:-200:+000:Weaponmastery
 R:k:+1000:+300:Archery
-R:k:+1000:+000:Sneakiness
-R:k:+500:+200:Disarming
+
```
Comment (BardurArantsson) [inline lib/mods/theme/edit/p_info.txt]: newlines

File: lib/mods/theme/edit/p_info.txt [hunk 6]
```diff
@@ -2477,7 +2381,7 @@ S:k:+0:+300:Conveyance
 S:k:+0:+300:Temporal
 S:Z:panic hit
 R:k:-1000:+000:Stealth
-R:k:-500:+000:Sneakiness
+
```
Comment (BardurArantsson) [inline lib/mods/theme/edit/p_info.txt]: newlines

File: src/cmd1.cc [hunk 7]
```diff
@@ -2859,7 +2746,7 @@ static bool_ easy_open_door(int y, int x)
 	else if (c_ptr->feat >= FEAT_DOOR_HEAD + 0x01)
 	{
 		/* Disarm factor */
-		i = p_ptr->skill_dis;
+		i = 100;
```
Comment (BardurArantsson) [inline src/cmd1.cc]: Is 100 the (theoretical) maximum?

File: src/cmd2.cc [hunk 8]
```diff
@@ -556,58 +545,9 @@ void do_cmd_go_down(void)
 
 		/* Leaving */
 		p_ptr->leaving = TRUE;
-
-		if (!fall_trap)
-		{
```
Comment (BardurArantsson) [inline src/cmd2.cc]: AFAICT this block should not be removed; "fall_trap" was false by default, and !false == true.

File: lib/help/debug.txt [hunk 9]
```diff
@@ -29,7 +29,7 @@ will not be scored if you use debug commands.
  *****debug.txt*29[o  Edit object attributes]               O  (unused)
  *****debug.txt*31[p  Phase door]                           P  (unused)
  *****debug.txt*33[q  Get a quest]                          Q  (unused)
- r  (unused)                             *****debug.txt*36[R  Create a trap]
```
Comment (BardurArantsson) [inline lib/help/debug.txt]: Isn't there also some mention of the (D)isarm command in the regular command listing?

File: src/cmd7.cc [hunk 10]
```diff
@@ -3555,7 +3554,7 @@ void do_cmd_unbeliever()
 	/* Select what to do */
 	while (TRUE)
 	{
-		if (!get_com("Disrupt [C]ontinuum or [D]etect Traps", &ch))
+		if (!get_com("Disrupt [C]ontinuum", &ch))
```
Comment (BardurArantsson) [inline src/cmd7.cc]: I think we should just ditch the whole menu here and promote the disruption of the "Anti-magic" skill... but that should probably be done as a follow-up.

File: lib/edit/k_info.txt [hunk 11]
```diff
@@ -2095,22 +2087,6 @@ W:5:0:5:0
...
 
-N:203:Trap Creation
-G:?:d
-I:70:7:0
-W:10:0:5:0
-A:10/1
-D:If you read this rather annoying scroll, snares and pitfalls will magically be planted all around
-D:you, ready to do nasty things to you once you walk onto them.
-
-N:204:Trap/Door Destruction
```
Comment (BardurArantsson) [inline lib/edit/k_info.txt]: I'm a bit worried about broken doors blocking areas off (esp. for Anti-magic users) unintentionally. AFAIK the only way to handle them is through door destruction type spells/effects. (I see that there's a GF_KILL_DOOR effect, so there's possibly some code somewhere which can do this and which doesn't handle traps.)

File: src/cmd7.cc [hunk 12]
```diff
@@ -3580,24 +3574,6 @@ void do_cmd_unbeliever()
...
-
-			if (skill < 25)
-			{
-				msg_print("You cannot use your detection abilities yet.");
-				break;
-			}
-
-			detect_traps(DEFAULT_RADIUS);
-
-			if (skill >= 35) destroy_doors_touch();
```
Comment (BardurArantsson) [inline src/cmd7.cc]: Ah, here it is. I don't your changes remove destroy_doors_touch(), right?

I think we should keep the "D" anti-magic ability and just rename it "[D]estroy door" or something.

File: src/dungeon.cc [hunk 13]
```diff
@@ -4185,12 +4158,6 @@ static void process_command(void)
 			do_cmd_macro_recorder();
 			break;
 		}
-	case CMD_BLUNDER:
-		{
-			if (do_control_walk()) break;
-			do_cmd_walk(always_pickup, FALSE);
```
Comment (BardurArantsson) [inline src/dungeon.cc]: I think the second parameter to do_cmd_walk can be removed as well.

File: src/generate.cc [hunk 14]
```diff
@@ -980,11 +939,11 @@ static void place_random_door(int y, int x)
 		cave_set_feat(y, x, FEAT_BROKEN);
 	}
 
-	/* Secret doors (200/1000) */
+	/* Secret doors (200/1000) - now locked instead */
 	else if (tmp < 600)
 	{
-		/* Create secret door */
-		place_secret_door(y, x);
```
Comment (BardurArantsson) [inline src/generate.cc]: Are secret doors on fixed levels, e.g. Death Watch, also handled correctly? (I.e. changed to locked doors.)

File: src/generate.cc [hunk 15]
```diff
@@ -4786,16 +4674,15 @@ static void fill_treasure(int x1, int x2, int y1, int y2, int difficulty)
 					}
 					else
 					{
-						place_trap(y, x);
+						/* Do nothing */
```
Comment (BardurArantsson) [inline src/generate.cc]: Just remove the "else { ... }" entirely?

File: src/generate.cc [hunk 16]
```diff
@@ -4815,8 +4702,7 @@ static void fill_treasure(int x1, int x2, int y1, int y2, int difficulty)
 				}
 				else if (value < 50)
 				{
-					/* Trap */
-					place_trap(y, x);
+					/* Do nothing */
```
Comment (BardurArantsson) [inline src/generate.cc]: I wonder what effect removing this if branch would have...

Soooo many ad-hoc magic numbers :(

File: src/init1.cc [hunk 17]
```diff
@@ -360,7 +358,7 @@ static cptr r_info_flags6[] =
 	"TELE_AWAY",
 	"TELE_LEVEL",
 	"DARKNESS",
-	"TRAPS",
+	"TRAPS",	/* Unused */
```
Comment (BardurArantsson) [inline src/init1.cc]: Just "XXX" it. That way any left-over "TRAPS" flags in the r_info.txt file will yield a parse error. (These should be trivial to remove automatically now with a "grep -v ^S:TRAPS" command, btw.)

File: src/init1.cc [hunk 18]
```diff
@@ -1956,12 +1916,9 @@ errr init_player_info_txt(FILE *fp)
 			if (8 != sscanf(buf + 4, "%d:%d:%d:%d:%d:%d:%d:%d",
 			                &s[0], &s[1], &s[2], &s[3], &s[4], &s[5], &s[6], &s[7])) return (1);
 
-			rp_ptr->r_dis = s[0];
```
Comment (BardurArantsson) [inline src/init1.cc]: This needs a follow-up at some point to actually remove these fields from the file entirely.

File: src/init1.cc [hunk 19]
```diff
@@ -930,44 +928,6 @@ static cptr d_info_flags2[] =
 	"XXX1"
 };
 
-/*
- * Trap flags
- */
-static cptr t_info_flags[] =
```
Comment (BardurArantsson) [inline src/init1.cc]: I think there may also be an array `k_info_flags2_trap` in this file which can now be removed.

File: src/init1.cc [hunk 20]
```diff
@@ -8136,7 +7911,7 @@ static errr process_dungeon_file_aux(char *buf, int *yval, int *xval, int xvalst
 				}
 				else
 				{
-					place_trap(y, x);
+					/* No traps - do nothing */
```
Comment (BardurArantsson) [inline src/init1.cc]: Remove the "else { ... }" entirely.

File: src/init1.cc [hunk 21]
```diff
@@ -8171,10 +7946,9 @@ static errr process_dungeon_file_aux(char *buf, int *yval, int *xval, int xvalst
 					place_object(y, x, TRUE, TRUE, OBJ_FOUND_SPECIAL);
 				}
 			}
-			/* Random trap */
 			else if (random & RANDOM_TRAP)
```
Comment (BardurArantsson) [inline src/init1.cc]: Remove RANDOM_TRAP define?

File: src/melee2.cc [hunk 22]
```diff
@@ -4238,7 +4237,6 @@ static bool_ make_attack_spell(int m_idx)
 				disturb(1);
 				if (blind) msg_format("%^s mumbles, and then cackles evilly.", m_name);
 				else msg_format("%^s casts a spell and cackles evilly.", m_name);
-				(void)trap_creation();
```
Comment (BardurArantsson) [inline src/melee2.cc]: The "TRAPS" spell should just be removed entirely.

File: src/melee2.cc [hunk 23]
```diff
@@ -6975,135 +6966,128 @@ static void process_monster(int m_idx, bool_ is_frien)
...
 			}
 
-			/* Check for monster trap */
-			if (c_ptr->feat == FEAT_MON_TRAP)
-			{
-				if (mon_hit_trap(m_idx)) return;
-			}
-			else
-			{
-				/* Copy list of objects; we need a copy because we're mutating the list. */
```
Comment (BardurArantsson) [inline src/melee2.cc]: Could would please restore indentation here? (By leaving the surrounding "{".) It's really hard to see what changed.

File: src/object1.cc [hunk 24]
```diff
@@ -3013,141 +2971,99 @@ bool_ object_out_desc(object_type *o_ptr, FILE *fff, bool_ trim_down, bool_ wait
 			text_out("It makes you invisible.  ");
 		}
 
-		if (o_ptr->tval != TV_TRAPKIT)
+		vn = 0;
+		if (f2 & (TR2_SUST_STR))
 		{
-			vn = 0;
```
Comment (BardurArantsson) [inline src/object1.cc]: Same thing here wrt. indentation. Just leave the surrounding "{" in place.

(I'd like the see the indentation removed eventually, but it can be done after everything is finalized and merged.)

File: src/q_eol.cc [hunk 25]
```diff
@@ -76,35 +75,6 @@ static bool_ quest_eol_gen_hook(void *, void *, void *)
 		done = generate_fracave(y0, x0, xsize, ysize, cutoff, FALSE, TRUE);
 	}
 
-	/* Place a few traps */
-	for (x = xsize - 1; x >= 2; x--)
```
Comment (BardurArantsson) [inline src/q_eol.cc]: Ah, here's your Eol problem. Note that this actually also places Eol himself in the level!

File: src/q_eol.cc [hunk 26]
```diff
@@ -76,35 +75,6 @@ static bool_ quest_eol_gen_hook(void *, void *, void *)
...
-				m_allow_special[r_idx] = TRUE;
-				m_idx = place_monster_one(y, x, r_idx, 0, FALSE, MSTATUS_ENEMY);
-				m_allow_special[r_idx] = FALSE;
-				// Mark with the QUEST flag
-				if (m_idx) m_list[m_idx].mflag |= MFLAG_QUEST;
-			}
-
-			if (magik(18))
-			{
-				place_trap(y, x);
```
Comment (BardurArantsson) [inline src/q_eol.cc]: This is the only bit that should be removed.

File: src/q_fireprof.cc [hunk 27]
```diff
@@ -476,84 +475,39 @@ static bool_ fireproof_gen_hook(void *, void *, void *)
 		return FALSE;
 	}
 
-	/* Go ahead */
+	/* load the map */
 	{
-		int traps, trap_y, trap_x;
-
```
Comment (BardurArantsson) [inline src/q_fireprof.cc]: You've also removed far too much code here.

File: src/skills_defs.hpp [hunk 28]
```diff
@@ -37,9 +37,9 @@
 #define SKILL_MIMICRY           32
 #define SKILL_ANTIMAGIC         33
 #define SKILL_RUNECRAFT         34
-#define SKILL_SNEAK             35
+/* XXX */
```
Comment (BardurArantsson) [inline src/skills_defs.hpp]: remove

File: src/skills_defs.hpp [hunk 29]
```diff
@@ -37,9 +37,9 @@
...
 #define SKILL_STEALTH           36
-#define SKILL_DISARMING         37
+/* XXX */
```
Comment (BardurArantsson) [inline src/skills_defs.hpp]: remove

File: src/spells2.cc [hunk 30]
```diff
@@ -6227,9 +6123,6 @@ bool_ passwall(int dir, bool_ safe)
 	/* Update the monsters */
 	p_ptr->update |= (PU_DISTANCE);
 
-	/* Redraw trap detection status */
-	p_ptr->redraw |= (PR_FRAME);
```
Comment (BardurArantsson) [inline src/spells2.cc]: I'm not sure this should be removed, i.e. I think the comment may be wrong. (Regardless it's harmless to leave it.)

This is also one of those things I'd like to eventually fix. All this fretting about exactly which bits of the screen to update is absolutely pointless on modern computers.

**Comment by BardurArantsson** [comment]
Just in case you haven't noticed the comment among all the other comments: https://github.com/tome2/tome2/pull/6/files#r57531484 solves your Eol problem, I think

**Comment by miramor** [comment]
Thanks and thanks, I'll get on this stuff shortly. Wish there were a way to fix those newline issues programatically...

File: src/cmd1.cc (hunk 7 above)
Comment (miramor) [inline src/cmd1.cc]: No; but the check is against rand_int(100).

File: src/init1.cc (hunk 17 above)
Comment (BardurArantsson) [inline src/init1.cc]: Just a little additional note: You may need to use "grep --text -v ^S:TRAPS [...]" on the Theme files. It seems grep thinks r_info.txt (Theme) is binary and chops off a huge portion of the file otherwise.

**Comment by miramor** [comment]
So, my apologies for dropping off the face of the Earth for two weeks. Real Life got very nasty very suddenly. I might have time for the next round of trap removal fixes at this point, but it's low on my priority list, and I will probably be busy with other stuff shortly.

IOW, chances are nothing will happen here for a while.

(Especially as things are out of sync again after your last push to master.)

**Comment by BardurArantsson** [comment]
No problem at all. RL happens sometimes.

The conflicts *should* mostly be pretty trivial; the bulk of the changes I made are entirely mechanical. So it should really just be a matter of accepting the change revisions from "master" and re-applying any deletions (which the compiler will mostly point out for you).

FWIW, it's unlikely that I'll make any further large scale changes during the short term.

File: src/cmd2.cc (hunk 8 above)
Comment (miramor) [inline src/cmd2.cc]: I'm pretty sure this is actually not a problem - the original 'if' check there seems to be redundant, and only would apply w.r.t. traps anyway. I'll see if there's any staircase bugginess, but the code looks okay.

File: lib/edit/k_info.txt (hunk 11 above)
Comment (miramor) [inline lib/edit/k_info.txt]: Bit late for that. Maybe it okay to tunnel through doors instead? Vanilla does that.

File: src/cmd7.cc (hunk 12 above)
Comment (miramor) [inline src/cmd7.cc]: Sure. If AM can disrupt arrow traps, why not locks...

File: src/dungeon.cc (hunk 13 above)
Comment (miramor) [inline src/dungeon.cc]: I'll attempt that. It will probably break things all over the place, but we'll see.

File: src/cmd2.cc (hunk 8 above)
Comment (miramor) [inline src/cmd2.cc]: Yeah, staircases work fine here. I *think* the logic was just confused due to the special trap case.

File: src/generate.cc (hunk 14 above)
Comment (miramor) [inline src/generate.cc]: They should be, but I will do some debug-mode checking on that.

File: src/generate.cc (hunk 15 above)
Comment (miramor) [inline src/generate.cc]: Okay done.

File: src/cmd2.cc (hunk 8 above)
Comment (BardurArantsson) [inline src/cmd2.cc]: Hm? Are you seeing actual *shafts* being generated? (Rather than just stairs)

AFAICT this is the *only* place in the code where create_up_shaft is ever set to TRUE.

File: lib/edit/k_info.txt (hunk 11 above)
Comment (BardurArantsson) [inline lib/edit/k_info.txt]: Yeah, tunneling is fine.

File: src/dungeon.cc (hunk 13 above)
Comment (BardurArantsson) [inline src/dungeon.cc]: Well, it's easy to check if it's ever called with anything but FALSE.

It's also fine to do this in a follow-up PR just to make it simpler to review.

File: src/generate.cc (hunk 14 above)
Comment (BardurArantsson) [inline src/generate.cc]: Edit: Deathwatch has granite walls now, but I'm not actually sure it ever had secret doors?

Yup, the two artifacts are hidden behind secret doors. I believe the doors to the stairs down is also a hidden door. (You need pretty searching to detect them, so "Slug-like" and "Coward" and hanging around for quite a few turns.)

Granite is fine.

> (Note: IMO we should change the default colors for permawalls and illusion walls.)

I think it's always been like this; even the secret doors were colored like granite (so different from permawalls).

Regardless, a different issue.

**Comment by BardurArantsson** [comment]
Ok, so I've pushed a few rather large flag_set<> refactors, so you're bound to get conflicts again. Sorry about that! (I really wanted to avoid 'sitting' on code for too long.)

From the perspective of this branch I *think* it should mostly be trivial-esque you-have-deleted-something-that-changed-in-your-upstream-branch, so hopefully it won't be too painful to resolve conflicts. If there's any conflicts you have trouble with, just let me know and I'll try to help as best I can.

**Comment by miramor** [comment]
Okay - I made a new branch off [rm-traps-v2](https://github.com/miramor/tome2/tree/rm-traps-v2), fixed the conflicts, merged master, and pushed it to [rm-traps-v3](https://github.com/miramor/tome2/tree/rm-traps-v3). It does not compile yet though. The build fails with an inscrutable template error on files.cc:

`
/home/proteus/tome2/src/files.cc:1924:121: error: ‘{anonymous}::detail::object_flag_cell_empty’ is not a valid template argument for type ‘const {anonymous}::object_flag_cell&’ because object ‘{anonymous}::detail::object_flag_cell_empty’ has not external linkage
 ct_flag_cell, detail::object_flag_cell_append, detail::object_flag_cell_empty>;
                                                                              ^
/home/proteus/tome2/src/files.cc: In function ‘void {anonymous}::display_flag_row(int, int, const std::__debug::vector<std::tuple<char, int, flag_set<7ul> > >&, const std::__debug::vector<const object_flag_meta*>&)’:
/home/proteus/tome2/src/files.cc:1996:13: error: ‘object_flag_cell_monoid’ has not been declared
  auto acc = object_flag_cell_monoid::empty;
             ^
/home/proteus/tome2/src/files.cc:2001:31: error: ‘object_flag_cell_monoid’ has not been declared
   object_flag_cell combined = object_flag_cell_monoid::empty;
                               ^
/home/proteus/tome2/src/files.cc:2013:17: error: ‘object_flag_cell_monoid’ has not been declared
      combined = object_flag_cell_monoid::append(
                 ^
/home/proteus/tome2/src/files.cc:2021:9: error: ‘object_flag_cell_monoid’ has not been declared
   acc = object_flag_cell_monoid::append(acc, combined);
         ^
`

Hoping you can make some sense out of that. Template metaprogramming confuses the dickens out of me.

**Comment by BardurArantsson** [comment]
Hm, that looks pretty odd. Does it work when you compile 'master' without any of your changes?

**Comment by miramor** [comment]
I believe it did, but let me check again.

**Comment by BardurArantsson** [comment]
Which compiler versions? (gcc/clang as appropriate)

**Comment by miramor** [comment]
```gcc (GCC) 5.3.1 20160406 (Red Hat 5.3.1-6)```

**Comment by miramor** [comment]
master actually compiles fine with clang.
```
clang version 3.7.0 (tags/RELEASE_370/final)
Target: x86_64-redhat-linux-gnu
Thread model: posix
```

Weird. Bug in one or the other compiler's C++11 support?

**Comment by BardurArantsson** [comment]
Sounds like it might be. (Probably the C++14 support, but meh.)

Does it work with gcc if you remove the "constexpr" in front of the line which declares `object_flag_cell_empty`? This is the line:

```
constexpr object_flag_cell object_flag_cell_empty { '\0', 0, nullptr };
```

**Comment by miramor** [comment]
That still borks with GCC.

```
In file included from /home/proteus/tome2/src/files.cc:26:
/home/proteus/tome2/src/monoid.hpp:27:21: error: constexpr variable 'empty' must be initialized by a constant expression
        static constexpr T empty = empty_;
                           ^       ~~~~~~
/home/proteus/tome2/src/files.cc:1996:13: note: in instantiation of template class 'monoid<(anonymous namespace)::object_flag_cell, &(anonymous
      namespace)::detail::object_flag_cell_append, &(anonymous namespace)::detail::object_flag_cell_empty>' requested here
        auto acc = object_flag_cell_monoid::empty;
                   ^
/home/proteus/tome2/src/monoid.hpp:27:29: note: read of non-constexpr variable 'object_flag_cell_empty' is not allowed in a constant expression
        static constexpr T empty = empty_;
                                   ^
/home/proteus/tome2/src/monoid.hpp:27:29: note: in call to 'object_flag_cell(object_flag_cell_empty)'
/home/proteus/tome2/src/files.cc:1920:18: note: declared here
object_flag_cell object_flag_cell_empty { '\0', 0, nullptr };
                 ^
```

and breaks it with Clang as well.

```
/home/proteus/tome2/src/monoid.hpp:27:21: error: constexpr variable 'empty' must be initialized by a constant expression
        static constexpr T empty = empty_;
                           ^       ~~~~~~
/home/proteus/tome2/src/files.cc:1996:13: note: in instantiation of template class 'monoid<(anonymous namespace)::object_flag_cell, &(anonymous
      namespace)::detail::object_flag_cell_append, &(anonymous namespace)::detail::object_flag_cell_empty>' requested here
        auto acc = object_flag_cell_monoid::empty;
                   ^
/home/proteus/tome2/src/monoid.hpp:27:29: note: read of non-constexpr variable 'object_flag_cell_empty' is not allowed in a constant expression
        static constexpr T empty = empty_;
                                   ^
/home/proteus/tome2/src/monoid.hpp:27:29: note: in call to 'object_flag_cell(object_flag_cell_empty)'
/home/proteus/tome2/src/files.cc:1920:18: note: declared here
object_flag_cell object_flag_cell_empty { '\0', 0, nullptr };
                 ^
```

**Comment by BardurArantsson** [comment]
Oh, right. It's times like this I really hate C++... Could you try removing the `constexpr` from line 27 in monoid.hpp too? Perhaps change it to "const" (which should be fine, regardless)

If that doesn't work, I'm giving up, at least for today, so you'll have to use `clang` until further notice :).

**Comment by miramor** [comment]
Those changes break for both compilers (again).

I'll stick with Clang for now.

Just my 2c though, we should prefer GCC over Clang when necessary, because GCC still far more ubiquitous. (And I don't think any major distro uses GCC 6 by default; in fact, Fedora 23 doesn't even make that version available in the repos.)

**Comment by BardurArantsson** [comment]
Ideally, I'd obviously want to compile on both gcc/clang (and I try to remember to test that from time to time).

Arch Linux has been on GCC 6.x for quite a while and I *think* it's a major distro(?!?). Fedora 24 (just released) has GCC 6.x.  Not sure what the current status is on Ubuntu/Debian, but obviously those are also major distros where we'd *ideally* want to be able to compile.

It's just quite problematic because I'm relying quite heavily on `constexpr` to make things a lot saner than they could possibly be without it...

**Comment by miramor** [comment]
So, just FYI: current master will compile even on Debian Stable, as long as you use Clang.

Also, I have to retract what I said about GCC vs. Clang support. The LLVM project provides a current Windows installer for Clang (including a fully 64-bit version); whereas MinGW is long obsolete, and MinGW-W64 is still somewhat out of date vs. GCC on Linux.

(It would still be better if Windows binaries could be cross-compiled on Linux, though. I will see today if I can get that working.)

**Comment by BardurArantsson** [comment]
That's something, at least :).

I think MSYS2 seems to be more popular these days? Maybe that's a viable option for cross-compiling.

**Comment by miramor** [comment]
Returning to this after two months of more important things... (External circumstances, the usual.)

It turns out the Clang versions of Windows are not designed to be used with MSYS and whatnot. They need headers from the Windows SDK, or perhaps Visual Studio, either of which are enormous downloads.

I don't think I'd heard of MSYS2 until today. But cross-builds from Linux would still be nicer IMO. Setting up a Linux VM or container takes all of 10 minutes; setting up a Win32 VM, let alone a 64-bit one, takes hours and gigabytes and too many headaches.

But, unfortunately I'm not sure cross-builds are possible right now. Cross-building on Linux with Clang via MinGW headers *might* be doable, but is definitely not supported by anyone upstream. It would be much better if GCC could be made to work dependably. And currently, I'm not even sure that the newer GCC versions work at all. (I just got myself stuck again on another absurd distro that doesn't ship them.)

In conclusion, I really think you should tag the last "pure" C revision, as I have here:

https://github.com/miramor/tome2/commits/pure-c

so that people can easily check it out and build it. Because most people will *not* be able to build the C++ ones.

**Comment by BardurArantsson** [comment]
Since I've been kind of annoying and pushed a couple of huge batches of things which conlict (albeit probably only in minor ways), I wonder if I should try to take over and get the last few bits of this in shape and merged.

Would you mind me doing that?

(As per usual I can't make any promises that it'll actually *happen* any time soon, just wanted to ask if you were OK with it. I'll obviously leave you as "author" in the git log :). )

**Comment by miramor** [comment]
I wouldn't mind at all. Please don't worry too much about authorship stuff in the git logs, either.

**Comment by BardurArantsson** [comment]
Great. I'll just leave this open as a reminder :)

**Comment by probonopd** [comment]
@BardurArantsson 
> Maybe the real solution here is to provide binaries or something like "flatpak" packages.
Given that you are already using Travis CI, you could [easily](https://github.com/search?utf8=%E2%9C%93&q=%22Package+the+binaries+built+on+Travis+CI+as+an+AppImage%22&type=Code&ref=searchresults) produce [AppImage](http://appimage.org/)...

**Comment by miramor** [comment]
@probonopd 

I think AppImage has issues running stuff on older systems that was compiled on newer ones. If the glibc versions differ, it probably won't run.

For general binary distribution purposes AppImage sounds okay, though I'm not sure it's great to have edit files and whatnot stuck in a read-only ISO filesystem.

**Comment by BardurArantsson** [comment]
Ok, so I have this on a private branch now and it's *almost* ready for merging+pushing, so I'll just close this.(Hopefully I'll get the bits ready this weekend.)

@probonopd We actually aren't using Travis CI, but a PR would probably be welcome :). If you're not up for a PR, if you could file a separate issue, that'd be great. (Can't promise when, if ever, I'll get to it, but y'know...)
